### ISSUE TITLE
[expo-web-browser] Clear authSession for expo-web-browser 

### ISSUE THREAD

**Issue by VladCherniavsky** [expo/expo#10982]
Hi all!

Seems as there is an issue with `expo-web-browser` dismissing session.
Currently I am using `await WebBrowser.openAuthSessionAsync(authUrl, redirectUrl);` to allow users to input their login credentials.  When user click 'sign out'  then `dismissAuthSession` is called. The problem is that user doesn't see  login form when he clicks login button again.

It looks like expo expo-web-browser saves token/cookies and uses them all the time.

Is the any way to clear those data  to provide ability for a user to log in into different accounts ?

**Comment by Csierram96** [comment]
any updates on this?

**Comment by iammorganparry** [comment]
Did this ever get resolved? Facing the same issue also

**Comment by keharv** [comment]
Would love for a solution on this topic.

**Comment by MoeKanan** [comment]
cc-ing @brentvatne for visibility. We're facing a similar issue with social logins. Users can't log in to another account since they get logged in automatically to the same account.

**Comment by piechoo** [comment]
I'm also having problem with this - how to clear sign in info ? Is there any option to clear browser cookies ?

**Comment by piechoo** [comment]
I found a way to achieve this on iOS  - you can pass option `preferEphemeralSession ` which will cause to use new session with every login request. Hovewer this would only solve problem on iOS, as far as i know there is no possible way to open chrome custom tab in incognito mode - so maybe there is some way to clear cookies or sth like that? This problem encounters f.ex when logging in with azure active directory - after logging out user should close the browser in order to log out completely - but how to achieve that using expo-web-browser on android?

**Comment by MoeKanan** [comment]
@VladCherniavsky were you able to find a fix?

**Comment by MoeKanan** [comment]
Thanks @brentvatne! Do you happen to know what is the parameter for AWS Cognito? I thought it was `prompt=select_account` but that doesn't work.

We use expo-auth-session and noticed that we can use `preferEphemeralSession` for iOS and it works great (although typescript doesn't like it because expo-auth-session doesn't include it)
links to reference https://docs.expo.dev/versions/latest/sdk/auth-session/ & https://docs.expo.dev/versions/latest/sdk/webbrowser/#authsessionopenoptions

However, we couldn't find a solution for Android. I read that Chrome custom tabs share the cookies with Chrome and attempting to remove cookies doesn't help much in react native.

**Comment by brentvatne** [comment]
i'm not familiar with cognito, but you might be able to find an answer on their issues or on stackoverflow, eg: https://stackoverflow.com/questions/58154256/aws-cognito-how-to-force-select-account-when-signing-in-with-google

**Comment by VladCherniavsky** [comment]
> @VladCherniavsky were you able to find a fix?

Unfortunately, I did not find a fix

**Comment by radelcom** [comment]
we are also in a similar position. we couldn't login with another user for any particular Social login after they already use it once and logout.

**Comment by Crare** [comment]
very odd this does not exist. Users should be able to log out on their apps without deleting the whole app.
also AuthSession.dismiss() gives error on android. I'm guessing it is using Webrowser.dismissAuthSession() or WebBrowser.dismissBrowser(). But none of these clear the browser-cookies.

**Comment by louisiscoding** [comment]
Still an issue. Using Cognito. the `revokeAsync` function seems not to be enough

**Comment by Crare** [comment]
If Firefox is default browser. then I can just clear app storage and data for testing purposes. (not feasible to ask for user to do it in production) But if they have Chrome browser as default, then you need to delete the whole app and reset clear storage from the chrome browser as well. Main culprit here seems to be the browser storing the login data for later use.

**Comment by ofito** [comment]
`PreferEphemeralSession` works great on iOS but is not a consistent solution across OSs. The only thing that works (and I find it hard to believe there isn't a better solution) is to use the logout URL as the main uri and our `/auth` as redirect uri. Screwed in our case that we wanted to use AuthSession for a simple authentication, the PKCE generation... but as an alternative it works.

**Comment by BlagojeV93** [comment]
For Azure using `preferEphemeralSession` with `WebBrowser.openAuthSessionAsync` instead of the standard `promptAsync` method from expo-auth session does the job. Same approach for android with the addition of `'&prompt=login` (which is the provided parameter to re-enforce authorization by Azure) to the authorization url successfully opens the browser every time and allows users to enter credentials!

**Comment by ganeshacharyaincture** [comment]
> For Azure using `preferEphemeralSession` with `WebBrowser.openAuthSessionAsync` instead of the standard `promptAsync` method from expo-auth session does the job. Same approach for android with the addition of `'&prompt=login` (which is the provided parameter to re-enforce authorization by Azure) to the authorization url successfully opens the browser every time and allows users to enter credentials!

For any one looking for hints it is below
`const result = await WebBrowser
              .openAuthSessionAsync(
                `${authorizeEndpoint}`,
                redirectTo,
                {
                  preferEphemeralSession: true
                }
              )`

**Comment by balbany** [comment]
> For Azure using `preferEphemeralSession` with `WebBrowser.openAuthSessionAsync` instead of the standard `promptAsync` method from expo-auth session does the job. Same approach for android with the addition of `'&prompt=login` (which is the provided parameter to re-enforce authorization by Azure) to the authorization url successfully opens the browser every time and allows users to enter credentials!

Further to the above, have had good success with the following for both ios and android (using the useAuthRequest hook):

```
const [request, response, promptAsync] = useAuthRequest(
    {
      clientId,
      redirectUri,
      scopes,
      extraParams: {
        prompt: 'login',
      },
    },
    discovery,
  );
```

**Comment by zztaghizade** [comment]
#state
#RECOMMENDED. Opaque value used to maintain state between the request and the callback. Typically, Cross-Site Request Forgery (CSRF, XSRF) mitigation is done by cryptographically binding the value of this parameter with a browser cookie.
from [openid documentation](https://openid.net/specs/openid-connect-core-1_0.html#AuthorizationRequest). Maybe try generating a random state or removing it from your authrequest?

**Comment by Ninokeep** [comment]
Someone found a solution ?

**Comment by zztaghizade** [comment]
The issue is not exactly with browser cookies, since expo web browser auth session uses chrome custom tabs, which inherently does not let cookies be stored. The issue is with the auth session, some auth servers actually store your session in the server side, and it is related to the id token, not refresh or access tokens. So even if you revoke the access and refresh tokens (aka using dismiss auth session), your session still stays open. What you need to do is `await WebBrowser.openAuthSessionAsync(endSessionEndpoint, redirecUrl)`, where you can provide the id token in the body as well to avoid redundant logout prompts. Then, you will be redirected to a page that says successfully logged out, that has a redirect link that takes the user back to the app.
If you don't want to open a logout page at all, you have to check whether your auth server allows it.
In summary, since revoking the tokens does not end your session server-side, when you open the login link, you see that you are still logged in, and to fix this you must open the end session endpoint in your browser instead of fetching the revocation endpoint.

**Comment by mroswald** [comment]
> The issue is not exactly with browser cookies, since expo web browser auth session uses chrome custom tabs, which inherently does not let cookies be stored. The issue is with the auth session, some auth servers actually store your session in the server side, and it is related to the id token, not refresh or access tokens.

Are you sure about this? How would the auth server recognize the user without cookies of any kind?

**Comment by zztaghizade** [comment]
> > The issue is not exactly with browser cookies, since expo web browser auth session uses chrome custom tabs, which inherently does not let cookies be stored. The issue is with the auth session, some auth servers actually store your session in the server side, and it is related to the id token, not refresh or access tokens.
> 
> Are you sure about this? How would the auth server recognize the user without cookies of any kind?

Yes. If you open site info -> cookies and site data, you will see 0 bytes stored. Auth servers don't usually store cookies for authentication, since it would not be secure. They usually store jwts in the session storage or local storage of your browser, they also store it in the database. Therefore, if you do not open the page to log out, the jwt remains in the session storage indicating that you are still logged in, even if you have revoked the access token and refresh token from the database. As I mentioned earlier, the jwt icludes the id token, by which the auth server knows that you have not physically logged out. Jwts contain their own expiration data and don't need to check back with the server for validity until they expire.

**Comment by mroswald** [comment]
Actually in our case it was a cookie. We found that using `setEphemeralBrowsingEnabled` is a solution for us, but it requires patching `expo-web-browser`.

**Comment by jstr** [comment]
> Actually in our case it was a cookie. We found that using `setEphemeralBrowsingEnabled` is a solution for us, but it requires patching `expo-web-browser`.

Are you able to share the details please?

**Comment by raulTeoSalinas** [comment]
5 years and this still being an issue.

**Comment by jstr** [comment]
Unfortunately this is an inherent issue on Android due to OS limitations. The best workaround is to use a prompt parameter on OAuth/auth initialisation, e.g. `prompt=log_in`, which forces a log in flow. If you're using an auth platform their docs should describe this.

**Comment by mroswald** [comment]
> Are you able to share the details please?

I added this as a patch in `node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt`, but didn't get the time to make a contribution

```
diff --git a/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt b/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt
index b0e105d..345995b 100644
--- a/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt
+++ b/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserModule.kt
@@ -3,8 +3,10 @@ package expo.modules.webbrowser
 import expo.modules.core.errors.CurrentActivityNotFoundException
 import android.content.Intent
 import android.net.Uri
+import android.os.Build
 import android.os.Bundle
 import android.text.TextUtils
+import android.util.Log
 import androidx.browser.customtabs.CustomTabColorSchemeParams
 import androidx.browser.customtabs.CustomTabsIntent
 import androidx.core.os.bundleOf
@@ -21,6 +23,8 @@ private const val PREFERRED_BROWSER_PACKAGE = "preferredBrowserPackage"
 private const val DEFAULT_BROWSER_PACKAGE = "defaultBrowserPackage"
 
 private const val MODULE_NAME = "ExpoWebBrowser"
+private const val TAG = "ExpoWebBrowser"
+private const val MIN_SDK_FOR_EPHEMERAL_SESSION = 36
 
 class WebBrowserModule : Module() {
   private val context
@@ -131,6 +135,26 @@ class WebBrowserModule : Module() {
       builder.setShareState(CustomTabsIntent.SHARE_STATE_ON)
     }
 
+    if (options.preferEphemeralSession) {
+      if (Build.VERSION.SDK_INT >= MIN_SDK_FOR_EPHEMERAL_SESSION) {
+        try {
+          // Try to call setEphemeralBrowsingEnabled using reflection
+          // This ensures backward compatibility with older androidx.browser versions
+          val setEphemeralBrowsingEnabledMethod = CustomTabsIntent.Builder::class.java.getMethod(
+            "setEphemeralBrowsingEnabled",
+            Boolean::class.javaPrimitiveType
+          )
+          setEphemeralBrowsingEnabledMethod.invoke(builder, true)
+        } catch (e: NoSuchMethodException) {
+          Log.w(TAG, "preferEphemeralSession requires androidx.browser:browser:1.9.0 or higher. The option will be ignored.")
+        } catch (e: Exception) {
+          Log.w(TAG, "Failed to set ephemeral session: ${e.message}")
+        }
+      } else {
+        Log.w(TAG, "preferEphemeralSession requires API level $MIN_SDK_FOR_EPHEMERAL_SESSION or higher (current: ${Build.VERSION.SDK_INT}). The option will be ignored.")
+      }
+    }
+
     return builder.build().apply {
       // We cannot use the builder's method enableUrlBarHiding, because there is
       // no corresponding disable method and some browsers enable it by default.
diff --git a/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserOptions.kt b/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserOptions.kt
index d8a6693..c2a340c 100644
--- a/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserOptions.kt
+++ b/node_modules/expo-web-browser/android/src/main/java/expo/modules/webbrowser/WebBrowserOptions.kt
@@ -11,5 +11,6 @@ internal data class OpenBrowserOptions(
   @Field var enableDefaultShareMenuItem: Boolean = false,
   @Field var enableBarCollapsing: Boolean = false,
   @Field var showInRecents: Boolean = false,
-  @Field(key = "createTask") var shouldCreateTask: Boolean = true
+  @Field(key = "createTask") var shouldCreateTask: Boolean = true,
+  @Field var preferEphemeralSession: Boolean = false
 ) : Record

```

It requires you to push some versions (`androidXBrowser` to 1.9.0+ and I think it was `compileSdkVersion` to 36). Using `prompt=log_in` was not a solution for us, it pops up the login but once the user goes through, keycloak errors out "There is another session active, ..." inside the webview, blocking the login.

**Comment by chirix33** [comment]
> 5 years and this still being an issue.

Shocked myself

**Comment by chirix33** [comment]
@mroswald You made a contribution yet? I think that'll be a good modification

**Comment by arapocket** [comment]
Seriously why is this not possible this is nuts

**Comment by mroswald** [comment]
> [@mroswald](https://github.com/mroswald) You made a contribution yet? I think that'll be a good modification

I tried running the expo project locally to prepare the contribution but had issues. So far I was not able to make the contribution due to that.

**Comment by chirix33** [comment]
> I tried running the expo project locally to prepare the contribution but had issues. So far I was not able to make the contribution due to that.

Got it

**Comment by mroswald** [comment]
Preparing the PR right now https://github.com/expo/expo/pull/42253

**Comment by chirix33** [comment]
Perfect. Seen it.
